pkg/cortex/modules.go

```diff
@@ -196,15 +196,29 @@ func (t *Cortex) initQuerier(cfg *Config) (serv services.Service, err error) {
 
 	queryable, engine := querier.New(cfg.Querier, t.distributor, t.storeQueryable, tombstonesLoader, prometheus.DefaultRegisterer)
 
-	t.api.RegisterQuerier(queryable, engine, t.distributor)
+	// if we are not configured for single binary mode then the querier needs to register its paths externally
+	registerExternally := cfg.Target != All
+	handler := t.api.RegisterQuerier(queryable, engine, t.distributor, registerExternally)
 
 	// Query frontend worker will only be started after all its dependencies are started, not here.
 	// Worker may also be nil, if not configured, which is OK.
-	worker, err := frontend.NewWorker(cfg.Worker, httpgrpc_server.NewServer(t.server.HTTPServer.Handler), util.Logger)
+	worker, err := frontend.NewWorker(cfg.Worker, httpgrpc_server.NewServer(handler), util.Logger)
 	if err != nil {
 		return
 	}
 
+	// single binary will be mysteriously unresponsive unless worker is working.  warn and configure here
+	if worker == nil && cfg.Target == All {
+		address := fmt.Sprintf("127.0.0.1:%d", cfg.Server.GRPCListenPort)
+		level.Warn(util.Logger).Log("msg", "Worker is nil in single binary mode.  Attempting automatic worker configuration.  If queries are unresponsive consider configuring the worker explicitly.", "address", address)
+		cfg.Worker.Address = address
+
+		worker, err = frontend.NewWorker(cfg.Worker, httpgrpc_server.NewServer(handler), util.Logger)
+		if err != nil {
+			return
+		}
+	}
```

Comment:
If the worker is not actively attempting to pull queries from the frontend they will just queue up in the frontend silently (there is a metric but no logs).  Once the queue is full it will start 429'ing but until then all queries just hang until client timeout.

With this forced attempt at configuring the worker the logs will show clearly there is an issue b/c the worker will begin logging failed grpc connections if it cannot connect.

Reply:
I understand this but the solution looks a bit overcomplicated to me. Can't we simply override cfg.Worker.Address before the first time we call frontend.NewWorker() in case the configured address is empty string and we're running in single binary mode?

Reply:
Good suggestion.  Fixed!